js/codemeta_generation.js

```diff
@@ -7,8 +7,21 @@
 
 "use strict";
 
+const CODEMETA_CONTEXT_URL = 'https://doi.org/10.5063/schema/codemeta-2.0';
+const CONTEXTS = {
+    [CODEMETA_CONTEXT_URL]: `${window.location.origin}/contexts/codemeta.jsonld`
```

Comment:
On https://codemeta.github.io/codemeta-generator/, the value of `window.location.origin` is `https://codemeta.github.io/` but the file would be at `https://codemeta.github.io/codemeta-generator/contexts/codemeta.jsonld`.

Instead, you should be able to load the `.jsonld` from directly from `jsonldCustomLoader`, parse it and return it

Reply:
I am not sure if it would be the right way to do it, as importing JSON files in plain JS does not seem standard yet (see https://www.freecodecamp.org/news/how-to-read-json-file-in-javascript/ and https://caniuse.com/mdn-javascript_statements_import_import_assertions_type_json). Unless you were thinking about [fetching](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch) the file locally; in this case, the loader will have to face sync issues. Were you thinking about transforming the context file into JS (encapsulating its content into a variable), or were you having something else in mind?

(Also, `window.location` on https://codemeta.github.io/codemeta-generator/ gives me `https://codemeta.github.io/codemeta-generator/`, isn't it what you have?)

Reply:
> Unless you were thinking about [fetching](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch) the file locally

Yes, that's what I meant

> in this case, the loader will have to face sync issues.

What do you mean?

> Were you thinking about transforming the context file into JS (encapsulating its content into a variable), or were you having something else in mind?

just `JSON.parse()`

> (Also, `window.location` on https://codemeta.github.io/codemeta-generator/ gives me `https://codemeta.github.io/codemeta-generator/`, isn't it what you have?)

yeah that works

Reply:
> > Unless you were thinking about [fetching](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch) the file locally
> 
> Yes, that's what I meant
> 
> > in this case, the loader will have to face sync issues.
> 
> What do you mean?
> 

I will have to make the `documentLoader` `async`, to handle the `Promise` returned by `fetch` and change the way I am using the `documentLoader` (not in argument as it is now, to be sure it has completed). If it is what you had in mind then I'll go for it :)

Reply:
Oh I just noticed this fetches the context every time the user re-generates the document (which includes switching between fields). That's probably where the latency is coming from.

Instead you should load the context as soon as the page load (with `Fetch`), and use that in the document loader

Reply:
In the end, I'm not using `fetch('/contexts/codemeta.jsonld')`, as I have verified it does make the same call as with  ```const CONTEXTS = { [CODEMETA_CONTEXT_URL]: `${window.location.origin}/contexts/codemeta.jsonld` }``` and I believe [`fetch()` also use `window.location` to resolve the URL ](https://stackoverflow.com/questions/36369082/relative-paths-with-fetch-in-javascript). (Also, using `fetch()` I think it could introduce bugs due to synchronization handling).

My mistake is to use `window.location.origin`, which was working for cypress tests but would not be working in the GitHub pages as you said. I fixed it in a separate commit.

I also initialized the lib with the `contextLoader` in the beginning, but here also, I verified and it does not change the calls. But it is not a bad thing to set it once and for all.

Reply:
What I am saying is that with the current implementation, every time we generate a codemeta document, the document loader sends an HTTP request to get the context.

This caused the delay in tests, and would make the generator slow for people who don't have a good internet connection.

Instead, the context should be loaded once when loading the page, and never fetched again until the user refreshes the page.

Reply:
Ah, I got it now! I think it would not have been an issue thanks to the cache, but I made the change. :)

Reply:
Ah, good point. I tested with a local HTTP server that probably doesn't send the right header and it might work on GitHub pages.